File: backend/ModelCodeManagement.Api/ModelCodeManagement.Api/Services/Impl/AuditLogService.cs
using System.Text.Json;
using Microsoft.AspNetCore.Http;
using ModelCodeManagement.Api.DTOs;
using ModelCodeManagement.Api.Entities;
using ModelCodeManagement.Api.Data;
using Microsoft.EntityFrameworkCore;

namespace ModelCodeManagement.Api.Services.Impl
{
    /// <summary>
    /// 审计日志服务实现
    /// </summary>
    public class AuditLogService : IAuditLogService
    {
        private readonly ApplicationDbContext _context;
        private readonly IBaseRepository<AuditLog> _auditLogRepository;
        private readonly IUserContextService _userContextService;
        private readonly IHttpContextAccessor _httpContextAccessor;
        private readonly ILogger<AuditLogService> _logger;

        public AuditLogService(
            ApplicationDbContext context,
            IBaseRepository<AuditLog> auditLogRepository,
            IUserContextService userContextService,
            IHttpContextAccessor httpContextAccessor,
            ILogger<AuditLogService> logger)
        {
            _context = context;
            _auditLogRepository = auditLogRepository;
            _userContextService = userContextService;
            _httpContextAccessor = httpContextAccessor;
            _logger = logger;
        }

        /// <summary>
        /// 记录审计日志
        /// </summary>
        public async Task LogAsync(AuditLog log)
        {
            try
            {
                // 补充用户信息
                if (log.UserId == 0 && _userContextService.TryGetCurrentUserId(out int userId))
                {
                    log.UserId = userId;
                }
                
                if (string.IsNullOrEmpty(log.Username))
                {
                    log.Username = _userContextService.GetCurrentUsername();
                }

                // 补充请求信息
                var httpContext = _httpContextAccessor.HttpContext;
                if (httpContext != null)
                {
                    log.IpAddress ??= httpContext.Connection.RemoteIpAddress?.ToString();
                    log.UserAgent ??= httpContext.Request.Headers["User-Agent"].FirstOrDefault();
                    log.RequestPath ??= httpContext.Request.Path;
                    log.HttpMethod ??= httpContext.Request.Method;
                }

                await _auditLogRepository.AddAsync(log);
                await _context.SaveChangesAsync();
            }
            catch (Exception ex)
            {
                // 记录日志失败不应影响业务操作
                _logger.LogError(ex, "记录审计日志失败 - Action: {Action}", log.Action);
            }
        }

        /// <summary>
        /// 记录操作日志
        /// </summary>
        public async Task LogActionAsync(string action, string description, string? entityType = null, int? entityId = null)
        {
            var log = new AuditLog
            {
                Action = action,
                Description = description,
                EntityType = entityType,
                EntityId = entityId,
                Result = "Success"
            };

            await LogAsync(log);
        }

        /// <summary>
        /// 记录数据变更日志
        /// </summary>
        public async Task LogChangeAsync<T>(string action, T? oldValue, T? newValue, string entityType, int entityId) where T : class
        {
            var log = new AuditLog
            {
                Action = action,
                EntityType = entityType,
                EntityId = entityId,
                Description = $"{action} {entityType} (ID: {entityId})",
                OldValue = oldValue != null ? JsonSerializer.Serialize(oldValue) : null,
                NewValue = newValue != null ? JsonSerializer.Serialize(newValue) : null,
                Result = "Success"
            };

            await LogAsync(log);
        }

        /// <summary>
        /// 记录错误日志
        /// </summary>
        public async Task LogErrorAsync(string action, string description, string errorMessage)
        {
            var log = new AuditLog
            {
                Action = action,
                Description = description,
                Result = "Failed",
                ErrorMessage = errorMessage
            };

            await LogAsync(log);
        }

        /// <summary>
        /// 分页查询审计日志
        /// </summary>
        public async Task<ApiResponse<PagedResult<AuditLog>>> GetPagedAsync(AuditLogQueryDto query)
        {
            try
            {
                // 验证分页参数
                if (query.PageIndex <= 0) query.PageIndex = 1;
                if (query.PageSize <= 0) query.PageSize = 20;
                if (query.PageSize > 100) query.PageSize = 100; // 限制最大页面大小
                
                var queryable = _context.AuditLogs.AsQueryable();

                // 应用过滤条件
                if (query.UserId.HasValue)
                    queryable = queryable.Where(l => l.UserId == query.UserId.Value);
                
                if (!string.IsNullOrEmpty(query.Username))
                    queryable = queryable.Where(l => l.Username.Contains(query.Username));
                
                if (!string.IsNullOrEmpty(query.Action))
                    queryable = queryable.Where(l => l.Action == query.Action);
                
                if (!string.IsNullOrEmpty(query.EntityType))
                    queryable = queryable.Where(l => l.EntityType == query.EntityType);
                
                if (!string.IsNullOrEmpty(query.Result))
                    queryable = queryable.Where(l => l.Result == query.Result);
                
                if (query.StartDate.HasValue)
                    queryable = queryable.Where(l => l.CreatedAt >= query.StartDate.Value);
                
                if (query.EndDate.HasValue)
                    queryable = queryable.Where(l => l.CreatedAt <= query.EndDate.Value);

                // 总记录数
                var totalCount = await queryable.CountAsync();

                // 排序和分页
                var items = await queryable
                    .OrderByDescending(l => l.CreatedAt)
                    .Skip((query.PageIndex - 1) * query.PageSize)
                    .Take(query.PageSize)
                    .ToListAsync();

                var result = new PagedResult<AuditLog>
                {
                    Items = items,
                    TotalCount = totalCount,
                    PageIndex = query.PageIndex,
                    PageSize = query.PageSize
                };

                _logger.LogInformation("分页查询审计日志成功 - 获取{Count}条记录，共{Total}条", items.Count, totalCount);
                return ApiResponse<PagedResult<AuditLog>>.SuccessResult(result);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "查询审计日志失败 - Query: {@Query}", query);
                return ApiResponse<PagedResult<AuditLog>>.ErrorResult($"查询审计日志失败: {ex.Message}");
            }
        }

        /// <summary>
        /// 按用户查询审计日志
        /// </summary>
        public async Task<ApiResponse<List<AuditLog>>> GetByUserAsync(int userId, DateTime? startDate = null, DateTime? endDate = null)
        {
            try
            {
                var queryable = _context.AuditLogs
                    .Where(l => l.UserId == userId);

                if (startDate.HasValue)
                    queryable = queryable.Where(l => l.CreatedAt >= startDate.Value);
                
                if (endDate.HasValue)
                    queryable = queryable.Where(l => l.CreatedAt <= endDate.Value);

                var logs = await queryable
                    .OrderByDescending(l => l.CreatedAt)
                    .ToListAsync();

                _logger.LogInformation("查询用户审计日志成功 - UserId: {UserId}, Count: {Count}", userId, logs.Count);
                return ApiResponse<List<AuditLog>>.SuccessResult(logs);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "查询用户日志失败 - UserId: {UserId}", userId);
                return ApiResponse<List<AuditLog>>.ErrorResult($"查询用户日志失败: {ex.Message}");
            }
        }

        /// <summary>
        /// 按实体查询审计日志
        /// </summary>
        public async Task<ApiResponse<List<AuditLog>>> GetByEntityAsync(string entityType, int entityId)
        {
            try
            {
                var logs = await _context.AuditLogs
                    .Where(l => l.EntityType == entityType && l.EntityId == entityId)
                    .OrderByDescending(l => l.CreatedAt)
                    .ToListAsync();

                _logger.LogInformation("查询实体审计日志成功 - EntityType: {EntityType}, EntityId: {EntityId}, Count: {Count}", 
                    entityType, entityId, logs.Count);
                return ApiResponse<List<AuditLog>>.SuccessResult(logs);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "查询实体日志失败 - EntityType: {EntityType}, EntityId: {EntityId}", entityType, entityId);
                return ApiResponse<List<AuditLog>>.ErrorResult($"查询实体日志失败: {ex.Message}");
            }
        }

        /// <summary>
        /// 清理过期日志
        /// </summary>
        public async Task<ApiResponse> CleanupOldLogsAsync(int daysToKeep = 90)
        {
            try
            {
                var cutoffDate = DateTime.Now.AddDays(-daysToKeep);
                var logsToDelete = await _context.AuditLogs
                    .Where(l => l.CreatedAt < cutoffDate)
                    .ToListAsync();

                if (logsToDelete.Any())
                {
                    _context.AuditLogs.RemoveRange(logsToDelete);
                    var deletedCount = await _context.SaveChangesAsync();

                    await LogActionAsync("CleanupLogs", $"清理了{deletedCount}条{daysToKeep}天前的日志");
                    _logger.LogInformation("清理过期审计日志成功 - 清理了{Count}条{Days}天前的日志", deletedCount, daysToKeep);
                    
                    return ApiResponse.SuccessResult($"成功清理{deletedCount}条过期日志");
                }
                
                _logger.LogInformation("清理过期审计日志完成 - 没有需要清理的{Days}天前的日志", daysToKeep);
                return ApiResponse.SuccessResult("没有需要清理的过期日志");
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "清理过期日志失败 - DaysToKeep: {DaysToKeep}", daysToKeep);
                return ApiResponse.ErrorResult($"清理日志失败: {ex.Message}");
            }
        }
    }
}

File: frontend/src/services/domain/audit-log.service.ts
// audit-log.service.ts - 审计日志服务
/**
 * 审计日志业务服务
 * 使用新的服务层架构：ApiClient + 直接后端数据访问
 * 基于后端AuditLogController提供的API接口
 */

import { apiClient } from '../api/client';
import type {
  AuditLog,
  ServiceResponse,
  PaginationParams,
  PagedResponse
} from '../../types/domain';

/**
 * 审计日志查询参数
 */
export interface AuditLogQuery extends PaginationParams {
  username?: string;
  action?: string;
  entityType?: string;
  result?: string;
  startDate?: string;
  endDate?: string;
  pageIndex?: number;
}

/**
 * 审计日志管理服务接口
 */
export interface IAuditLogService {
  getAuditLogs(query: AuditLogQuery): Promise<PagedResponse<AuditLog>>;
  cleanupOldAuditLogs(days: number): Promise<ServiceResponse<boolean>>;
}

/**
 * 审计日志管理服务实现
 */
export class AuditLogService implements IAuditLogService {

  /**
   * 获取审计日志列表
   */
  async getAuditLogs(query: AuditLogQuery): Promise<PagedResponse<AuditLog>> {
    try {
      
      const params = new URLSearchParams();
      
      // 分页参数
      if (query.pageIndex !== undefined) {
        params.append('pageIndex', query.pageIndex.toString());
      } else if (query.page) {
        params.append('pageIndex', (query.page - 1).toString());
      }
      
      params.append('pageSize', (query.pageSize || 20).toString());
      
      // 筛选参数
      if (query.username?.trim()) {
        params.append('username', query.username.trim());
      }
      if (query.action?.trim()) {
        params.append('action', query.action.trim());
      }
      if (query.entityType?.trim()) {
        params.append('entityType', query.entityType.trim());
      }
      if (query.result?.trim()) {
        params.append('result', query.result.trim());
      }
      if (query.startDate?.trim()) {
        params.append('startDate', query.startDate.trim());
      }
      if (query.endDate?.trim()) {
        params.append('endDate', query.endDate.trim());
      }

      const response = await apiClient.get(`/audit-logs?${params.toString()}`);
      
      if (!response.success) {
        console.error('❌ [AuditLogService] API请求失败:', response.error);
        return {
          success: false,
          error: response.error || '获取审计日志失败'
        };
      }

      // 直接使用后端分页数据格式
      const backendData = response.data;
      const items = backendData?.items || backendData?.Items || [];
      
      // 直接转换为前端AuditLog格式
      const auditLogs: AuditLog[] = items.map((item: any) => ({
        id: item.Id?.toString() || item.id?.toString() || '',
        userId: item.UserId?.toString() || item.userId?.toString() || '',
        username: item.UserName || item.userName || '',
        action: item.Action || item.action || '',
        entityType: item.EntityType || item.entityType || '',
        entityId: item.EntityId?.toString() || item.entityId?.toString() || '',
        description: item.Description || item.description || '',
        result: item.Result || item.result || '',
        oldValue: item.OldValue || item.oldValue || '',
        newValue: item.NewValue || item.newValue || '',
        ipAddress: item.IpAddress || item.ipAddress || '',
        requestPath: item.RequestPath || item.requestPath || '',
        userAgent: item.UserAgent || item.userAgent || '',
        errorMessage: item.ErrorMessage || item.errorMessage || '',
        durationMs: item.DurationMs || item.durationMs || 0,
        createdAt: item.CreatedAt || item.createdAt || '',
        updatedAt: item.UpdatedAt || item.updatedAt || ''
      }));
      
      return {
        success: true,
        data: auditLogs,
        totalCount: backendData?.totalCount || backendData?.TotalCount || 0,
        currentPage: (backendData?.currentPage || backendData?.CurrentPage || query.pageIndex || 0) + 1,
        pageSize: backendData?.pageSize || backendData?.PageSize || query.pageSize || 20,
        totalPages: backendData?.totalPages || backendData?.TotalPages || 
          Math.ceil((backendData?.totalCount || backendData?.TotalCount || 0) / (query.pageSize || 20)),
        message: '获取成功'
      };
      
    } catch (error) {
      console.error('❌ [AuditLogService] getAuditLogs error:', error);
      return {
        success: false,
        error: `获取审计日志列表失败: ${error instanceof Error ? error.message : String(error)}`
      };
    }
  }

  /**
   * 清理过期审计日志
   */
  async cleanupOldAuditLogs(days: number): Promise<ServiceResponse<boolean>> {
    try {
      
      if (days <= 0) {
        return {
          success: false,
          error: '保留天数必须大于0'
        };
      }

      const response = await apiClient.delete(`/audit-logs/cleanup?days=${days}`);
      
      if (!response.success) {
        console.error('❌ [AuditLogService] API请求失败:', response.error);
        return {
          success: false,
          error: response.error || '清理审计日志失败'
        };
      }

      return {
        success: true,
        data: true,
        message: '清理成功'
      };
      
    } catch (error) {
      console.error('❌ [AuditLogService] cleanupOldAuditLogs error:', error);
      return {
        success: false,
        error: `清理审计日志失败: ${error instanceof Error ? error.message : String(error)}`
      };
    }
  }
}

// 创建默认实例
export const auditLogService = new AuditLogService();